Regression generators

smastelini opened the pull request:
Adds the following new regression synthetic data generators:

- FriedmanDrift, as described in this [paper](https://link.springer.com/article/10.1007/s10618-010-0201-y) (the one that introduces the well-known FIMT-DD tree regressor). It implements three variants of concept drift
- Planes2D (originally 2DPlanes)
- Mv

The last two generators are described in the [Luís Torgo](https://www.dcc.fc.up.pt/~ltorgo/Regression/DataSets.html) repository of regression datasets and were used in multiple research papers in the last years.

Moreover, this PR also modifies `river.synth.Friedman` by adding a gaussian noise to the label, as originally proposed in the [original paper](https://www.jstor.org/stable/2241837?seq=1).

jacobmontiel commented on river/datasets/synth/friedman.py:
```diff
@@ -51,6 +52,230 @@ def __iter__(self):
+        points covered by this variant, and during a window of length `transition_window`,
+        examples from both old and the new concepts are generated with equal probability.
+        After the transition period, only the examples from the new concept are generated.
+    change_point1
```
This can be a single parameter:
```python
position: tuple = (point_1, point_2, point_3)
```

smastelini replied:
This simplifies the usage for the user, indeed. I changed how the concept drift locations are set in commit [afa307d](https://github.com/online-ml/river/pull/387/commits/afa307da973ab1519b2d06ae3852ba566f5ee147), as suggested.

jacobmontiel commented on river/datasets/synth/friedman.py:
```diff
@@ -51,6 +52,230 @@ def __iter__(self):
             yield x, y
+
+
+class FriedmanDrift(base.SyntheticDataset):
```
Should this class extend the `Friedman` class?

smastelini replied:
Well, apart from clearly stating that `FriedmanDrift` is an extension of `Friedman`, there are no immediate gains. However, I think this is enough reason by itself.

Changed in commit commit [afa307d](https://github.com/online-ml/river/pull/387/commits/afa307da973ab1519b2d06ae3852ba566f5ee147).

jacobmontiel commented on river/datasets/synth/planes_2d.py:
```diff
@@ -0,0 +1,73 @@
+
+    $$P(x_m = -1) = P(x_m = 0) = P(x_m = 1) = \\frac{1}{3}, m=2,\\ldots, 10$$
+
+    The target value is defined by using rule:
```
Suggested rewording: "defined by the following rule"

smastelini replied:
Reworded in commit [afa307d](https://github.com/online-ml/river/pull/387/commits/afa307da973ab1519b2d06ae3852ba566f5ee147).

jacobmontiel reviewed (commented):
The implementation looks good to me.

There are just minor comments. Let me know if you need further information.

jacobmontiel commented:
It might be worth exploring the possibility of expanding the `ConceptDriftStream` generator to account for the kinds of drift in the `FriedmandDrift` generator. This is outside of the scope of this PR but we could create a new issue for this purpose. Comments @smastelini ?

smastelini commented:
Thanks for the review, @jacobmontiel! I addressed your comments in the last commit.

> It might be worth exploring the possibility of expanding the ConceptDriftStream generator to account for the kinds of drift in the FriedmandDrift generator. This is outside of the scope of this PR but we could create a new issue for this purpose. Comments @smastelini?

I think they are interesting cases to explore in the future. I am just not sure how straightforward it would be to apply these cases for generic data generators. FriedmanDrift was specifically crafted for regression tasks and the Friedman synthetic case. For sure, we can create an issue to evaluate the applicability of these drift variants to other tasks.

Please, let me know if you need more changes :)

jacobmontiel commented:
Thanks for the latest changes @smastelini 

The implementation is correct.
